File: trainer/src/train.js
const tf = require("@tensorflow/tfjs-node");
const fs = require("fs");

// Load Preprocessed Data
const rawData = JSON.parse(fs.readFileSync("./data/preprocessed.json"));

// Extract Features (X) and Labels (Y)
const features = rawData.map((row) => [
	row.DurationDays,
	row.ComplexityScore,
	row.CourtLevel,
	row.Region,
	row.ClaimAmount,
	row.SettlementAmount,
	row.LegalRepresentation,
	row.EvidenceScore,
	row.SettlementType,
]);

const labels = rawData.map((row) => row.RiskLevel);

// Convert to TensorFlow tensors
const xs = tf.tensor2d(features);
const ys = tf.tensor1d(labels, "int32");

// Define the model
const model = tf.sequential();
model.add(
	tf.layers.dense({
		inputShape: [features[0].length],
		units: 16,
		activation: "relu",
	})
);
model.add(tf.layers.dense({ units: 8, activation: "relu" }));
model.add(tf.layers.dense({ units: 3, activation: "softmax" }));

model.compile({
	optimizer: "adam",
	loss: "sparseCategoricalCrossentropy",
	metrics: ["accuracy"],
});

// Train the model
async function trainModel() {
	await model.fit(xs, ys, {
		epochs: 150,
		batchSize: 10,
		shuffle: true,
	});

	// Save the model
	await model.save("file://./models/legal_model");
	console.log("Model training complete and saved.");
}

trainModel();

File: trainer/src/train.py
import os
import tensorflow as tf
import json
import numpy as np

# Ensure the 'models' directory exists
os.makedirs('./models', exist_ok=True)

# Load Preprocessed Data
with open('./data/preprocessed.json', 'r') as file:
    raw_data = json.load(file)

# Extract Features (X) and Labels (Y)
features = np.array([[   
    row['DurationDays'], 
    row['ComplexityScore'], 
    row['CourtLevel'], 
    row['Region'], 
    row['ClaimAmount'], 
    row['SettlementAmount'], 
    row['LegalRepresentation'], 
    row['EvidenceScore'], 
    row['SettlementType']
] for row in raw_data])

labels = np.array([row['RiskLevel'] for row in raw_data])

# Define the model
model = tf.keras.Sequential([
    tf.keras.layers.Dense(16, activation='relu', input_shape=(features.shape[1],)),
    tf.keras.layers.Dense(8, activation='relu'),
    tf.keras.layers.Dense(3, activation='softmax')
])

model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

# Train the model
model.fit(
    features, 
    labels,
    epochs=150,
    batch_size=10,
    shuffle=True
)

# Save the model (now the directory exists)
model.save('./models/legal_model.keras')

print("✅ Model training complete and saved successfully!")
